pkg/apis/ceph.rook.io/v1/validate_resources_test.go

```diff
@@ -0,0 +1,127 @@
+package v1
+
+import (
+	"github.com/ghodss/yaml"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestCephClusterValidateCreate(t *testing.T) {
+	cephClusterYaml := getCephClusterCRD()
+	c := &CephCluster{}
+	err := yaml.Unmarshal(cephClusterYaml, c)
+	assert.Nil(t, err)
+	err = c.ValidateCreate()
+	assert.Nil(t, err)
+
+	c.Spec.External.Enable = true
+	c.Spec.Monitoring = MonitoringSpec{}
+	err = c.ValidateCreate()
+	assert.NotNil(t, err)
+}
+
+func TestValidatePoolSpec(t *testing.T) {
+	poolYaml := getCephBlockPoolCRD()
+	p := &CephBlockPool{}
+	err := yaml.Unmarshal(poolYaml, p)
+	assert.Nil(t, err)
+	err = ValidatePoolSpecs(p.Spec)
+	assert.Nil(t, err)
+
+	p.Spec.ErasureCoded.DataChunks = 1
+	err = ValidatePoolSpecs(p.Spec)
+	assert.NotNil(t, err)
+}
+
+func TestCephBlockPoolValidateUpdate(t *testing.T) {
+	poolYaml := getCephBlockPoolCRD()
+	p := &CephBlockPool{}
+	err := yaml.Unmarshal(poolYaml, p)
+	assert.Nil(t, err)
+
+	up := p.DeepCopy()
+	up.Spec.Replicated.Size = 3
+	up.Spec.Replicated.RequireSafeReplicaSize = true
+	err = up.ValidateUpdate(p)
+	assert.NotNil(t, err)
+}
+
+func TestCephClusterValidateUpdate(t *testing.T) {
+	cephClusterYaml := getCephClusterCRD()
+	c := &CephCluster{}
+	err := yaml.Unmarshal(cephClusterYaml, c)
+	assert.Nil(t, err)
+	err = c.ValidateCreate()
+	assert.Nil(t, err)
+
+	// Updating the CRD specs with invalid values
+	uc := c.DeepCopy()
+	uc.Spec.DataDirHostPath = "var/rook"
+	err = uc.ValidateUpdate(c)
+	assert.NotNil(t, err)
+}
+
+func getCephBlockPoolCRD() []byte {
+	return []byte(`
+apiVersion: ceph.rook.io/v1
+kind: CephBlockPool
+metadata:
+  name: ec-pool
+  namespace: rook-ceph
+spec:
+  failureDomain: osd
+  erasureCoded:
+    dataChunks: 2
+    codingChunks: 1
+  parameters:
+    compression_mode: none
+  annotations:
+  #  key: value
+`)
+}
+func getCephClusterCRD() []byte {
```

Comment:
Instead of creating yaml and unmarshalling it in the test, let's just use the golang types directly. For example, you should be able to directly declare any of the structs in types.go, change the properties, and then validate them.

Reply:
Done